### PR TITLE
PHOTOMOD PRF file format driver

### PR DESCRIPTION
**Pull request by drons**
GDAL driver for read-only access to tiled file format, produced by Racurs PHOTOMOD software (http://www.racurs.ru/index.php).

### CODE REVIEW

File: gdal/frmts/prf/phprfdataset.cpp [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+               VRTSourcedRasterBand( poDataset, nBandCount, eType, nXSize, nYSize )
+    {
+    }
+
+    void AddOverview( GDALRasterBand* ov )
+    {
+        osOverview.push_back( ov );
+    }
+
+    int GetOverviewCount()
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Tag it with override

File: gdal/frmts/prf/phprfdataset.cpp [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+        {
+            return (int)osOverview.size();
+        }
+        else
+        {
+            return VRTSourcedRasterBand::GetOverviewCount();
+        }
+    }
+
+    GDALRasterBand* GetOverview( int i )
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Tag it with override

File: gdal/frmts/prf/phprfdataset.cpp [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+    void AddOverview( GDALRasterBand* ov )
+    {
+        osOverview.push_back( ov );
+    }
+
+    int GetOverviewCount()
+    {
+        if( osOverview.size() > 0 )
+        {
+            return (int)osOverview.size();
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Please use C++ casts

File: gdal/frmts/prf/phprfdataset.cpp [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+        }
+        else
+        {
+            return VRTSourcedRasterBand::GetOverviewCount();
+        }
+    }
+
+    GDALRasterBand* GetOverview( int i )
+    {
+        size_t	n = (size_t)i;
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Please use C++ cast

File: gdal/frmts/prf/phprfdataset.cpp [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+        }
+    }
+
+    if( eResultDatatype == GDT_Unknown )
+    {
+        CPLError( CE_Failure, CPLE_OpenFailed, "GDAL Dataset datatype not found\n" );
+        return NULL;
+    }
+
+    if( nSizeX == 0 || nSizeY == 0 || nBandCount == 0 )
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Also test against negative values

File: gdal/frmts/prf/phprfdataset.cpp [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+        {
+            poBand->AddOverview( poTileBand );
+        }
+    }
+    return 0;
+}
+
+int PhPrfDataset::Identify( GDALOpenInfo* poOpenInfo )
+{
+    if( EQUAL( CPLGetExtension( poOpenInfo->pszFilename ), PH_PRF_EXT ) )
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Identification by extension only is fragile. As I see later that there must be a 'phini' root node, perhaps checking for this string in poOpenInfo->pabyHeader (test that it is not NULL) would make identification more robust

File: gdal/frmts/prf/phprfdataset.cpp [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+        PhPrfBand*	poBand = new PhPrfBand( this, i + 1, eType, nSizeX, nSizeY );
+        SetBand( i + 1, poBand );
+    }
+}
+
+PhPrfDataset::~PhPrfDataset()
+{
+}
+
+int PhPrfDataset::AddTile( const char* pszPartName, GDALAccess eAccessType, int nWidth, int nHeight, int nOffsetX, int nOffsetY, int nScale )
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Use bool return instead of 0 / -1 ?

File: gdal/frmts/prf/phprfdataset.cpp [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+        {
+            return psAttr->psChild->pszValue;
+        }
+    }
+    return osDef;
+}
+
+static bool ParseGeoref( const CPLXMLNode* psGeorefElt, double* padfGeoTrans )
+{
+    bool                pabOk[6] = {false,false,false,false,false,false};
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Naming convention. Should be abOk since the variable here is the array and not a pointer to the array. Same for papszGeoKeys

File: gdal/frmts/prf/phprfdataset.cpp [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+    static const char*  papszGeoKeys[6] = { "A_0", "A_1", "A_2",
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Should be static const char* const

File: gdal/frmts/prf/phprfdataset.cpp [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+    }
+
+    int             nSizeX = 0;
+    int             nSizeY = 0;
+    int             nBandCount = 0;
+    GDALDataType    eResultDatatype = GDT_Unknown;
+    CPLString       osPartsBasePath( CPLGetPath( poOpenInfo->pszFilename ) );
+    CPLString       osPartsPath( osPartsBasePath + "/" + CPLGetBasename( poOpenInfo->pszFilename ) );
+    CPLString       osPartsExt;
+    double          padfGeoTrans[6] = {0,0,0,0,0,0};
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: Naming convention: adfGeoTrans

File: gdal/frmts/prf/phprfdataset.cpp [hunk 11]
```diff
@@ -55,21 +55,21 @@ class PhPrfBand : public VRTSourcedRasterBand
         osOverview.push_back( ov );
     }
 
-    int GetOverviewCount()
+    int GetOverviewCount() CPL_OVERRIDE
```
Comment (rouault) [inline gdal/frmts/prf/phprfdataset.cpp]: you can just use override, even in non C++11 mode (in that case it will be an empty #define)

**Comment by rouault** [comment]
You'd better rebase on top of trunk as soon as https://github.com/OSGeo/gdal is (mostly) green

**Comment by drons** [comment]
@rouault please take a look at last commit 1cbe093.
It fixes dataset leak, but it is strange behavior of VRTDataset for me.

**Comment by drons** [comment]
Did you mean those changeset https://github.com/drons/gdal/commit/705ccd3456a68dfaafc3657ca6635e4aca64e24a?
It does not work. Subtile datasets still leak.

**Comment by rouault** [comment]
> Did you mean those changeset drons@705ccd3?

Yes, I've checkout this, and apparently the issue is when the tiles are VRT files. References and VRT are really messy. Anyway the current state of your branch looks good to me. Anything else left on your side ?

**Comment by drons** [comment]
I'm think that is all

**Comment by rouault** [comment]
Committed in r36731

**Comment by drons** [comment]
Thanks!
